nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,447 @@
+        sled: &'a Sled,
+        handle: ExtractedArtifactDataHandle,
+        hash: ArtifactHash,
+    },
+    CopyFromDepot {
+        target_sled: &'a Sled,
+        depot_sled: &'a Sled,
+        hash: ArtifactHash,
+    },
+    Delete {
+        sled: &'a Sled,
+        hash: String,
+    },
+}
+
+impl BackgroundTask for ArtifactReplication {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            match self.activate_internal(opctx).await {
+                Ok(status) => serde_json::to_value(status).unwrap(),
+                Err(err) => {
+                    let err_string = format!("{:#}", err);
+                    error!(
+                        &opctx.log,
+                        "error during artifact replication";
+                        "error" => &err_string,
+                    );
+                    serde_json::json!({
+                        "error": err_string,
+                    })
+                }
+            }
+        }
+        .boxed()
+    }
+}
+
+impl ArtifactReplication {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        local_rx: mpsc::Receiver<ArtifactsWithPlan>,
+    ) -> ArtifactReplication {
+        ArtifactReplication { datastore, local: Vec::new(), local_rx }
+    }
+
+    async fn activate_internal<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> anyhow::Result<TufArtifactReplicationStatus> {
+        let log = &opctx.log;
+        let datastore = &self.datastore;
+
+        // Move any received artifacts out of `local_rx` into `local`.
+        loop {
+            match self.local_rx.try_recv() {
+                Ok(artifacts) => self.local.push(artifacts),
+                Err(TryRecvError::Empty) => break,
+                Err(TryRecvError::Disconnected) => {
+                    // If the last sender for this channel is dropped, then the
+                    // `Nexus` type has also been dropped. This is presumably
+                    // a bug.
+                    panic!("artifact replication receiver disconnected");
+                }
+            }
+        }
+
+        // Query the database for artifacts that we ought to have.
+        let mut artifacts = HashMap::new();
+        let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+        while let Some(p) = paginator.next() {
+            let batch = datastore
+                .update_tuf_artifact_list(opctx, &p.current_pagparams())
+                .await?;
+            paginator = p.found_batch(&batch, &|a| a.sha256);
+            for artifact in batch {
+                artifacts
+                    .insert(artifact.sha256.0, ArtifactPresence::default());
+            }
+        }
+        // Ask all sled agents to list the artifacts they have, and mark those
+        // artifacts as present on those sleds.
+        let sleds = datastore
+            .sled_list_all_batched(&opctx, SledFilter::InService)
```

Comment:
Strong :+1: on adding a new variant to `SledFilter` here, specifically because it's not obvious whether "in service" or "commissioned" is correct here IMO. I think `Commissioned` is not right, because that would include sleds that have been expunged but not yet decommissioned, and (presumably) there's no reason to try to ask such a sled anything. 

(See also the languished #5995 where I tried to remove `InService` altogether in favor of more specific variants, although it seemed a little dubious in practice.)